src/stan/variational/advi.hpp

```diff
@@ -150,18 +150,143 @@ namespace stan {
+          eta_sequence.pop();
+
+          for (iter_tune = 1; iter_tune <= tuning_iter; ++iter_tune) {
+            m = (eta_sequence_size - eta_sequence.size() - 1) *
+              tuning_iter + iter_tune; // # of total tuning iterations
+            stan::services::variational::print_progress(
+              m, 0, tuning_iter*eta_sequence_size,
+              tuning_iter, true, "", "", *print_stream_);
+
+            // Compute gradient of ELBO
+            calc_ELBO_grad(variational, elbo_grad);
+
+            // Update learning rate parameters
+            if (iter_tune == 1) {
+              params_prop += elbo_grad.square();
+            } else {
+              params_prop = pre_factor * params_prop +
+                            post_factor * elbo_grad.square();
+            }
+            eta_scaled = eta / sqrt(static_cast<double>(iter_tune));
+
+            // Stochastic gradient update
+            variational += eta_scaled * elbo_grad / (tau + params_prop.sqrt());
+          }
+          elbo = calc_ELBO(variational);
+
+          // Check if:
+          // (1) ELBO at current eta is worse than the best ELBO
+          // (2) the best ELBO hasn't actually diverged
+          if (elbo < elbo_best && elbo_best > elbo_init) {
+            if (print_stream_) {
+              *print_stream_ << "Success!"
+                << " Found best tuned hyperparameter"
+                << " [eta = " << eta_best
+                << "]";
+              if (eta_sequence.size() > 0) {
+                *print_stream_
+                  << " earlier than expected.";
+              }
+              *print_stream_ << std::endl << std::endl;
+            }
+            do_more_tuning = false;
+          } else {
+            if (eta_sequence.size() > 0) {
+              // Restart phase
+              elbo_best = elbo;
+              eta_best = eta;
+            } else {
+              // No more eta values to try, so use current eta if it
+              // didn't diverge or fail if it did diverge
+              if (elbo > elbo_init) {
+                if (print_stream_)
+                  *print_stream_ << "Success!"
+                    << " Found best tuned hyperparameter"
+                    << " [eta = " << eta_best
+                    << "]" << std::endl << std::endl;
+                eta_best = eta;
+                do_more_tuning = false;
+              } else {
+                const char* name = "All proposed step-sizes";
+                const char* msg1 = "failed. Your model may be either "
+                  "severely ill-conditioned or misspecified.";
+                stan::math::domain_error(function, name, "", msg1);
+                return 0;
+              }
+            }
+            // Reset
+            params_prop = Q(model_.num_params_r());
+          }
+          variational = Q(cont_params_);
+        }
+        return eta_best;
+      }
+
+      /**
+       * Runs stochastic gradient ascent with an adaptive stepsize.
```

Comment:
using the `break` paradigm is messier because we need to reset the variational distribution Q before trying an element of `eta_sequence` and before returning `eta_best`.

a queue made sense to me because we might want to chance `eta_sequence` down the road (add/remove values) and it makes sense that we are working our way down a queue in the procedure. i could change this if you feel strongly though.

i've implemented the other changes you suggested.